### PR TITLE
Remove IResponseHandler in C#

### PR DESCRIPTION
**Pull request by bernardnormier**
This is a small follow-up to my previous PR. It removes IResponseHandler (a callback interface) by switching to await calls in Connection and CollocatedRequestHandler.

It also removes for now support for incoming batch requests, as we currently can't test this feature. Support for sending batch requests was already removed some time ago.

### CODE REVIEW

File: csharp/src/Ice/Protocol.cs [hunk 1]
```diff
@@ -158,6 +159,45 @@ public static Ice.OutputStream CreateFailureResponseFrame(System.Exception excep
...
+            }
+            else
+            {
+                facet = "";
+            }
+
+            string operation = requestFrame.ReadString();
+            byte mode = requestFrame.ReadByte();
+            var context = new Dictionary<string, string>();
+            int sz = requestFrame.ReadSize();
```
Comment (pepone) [inline csharp/src/Ice/Protocol.cs]: Shouldn't this code use `var context = requestFrame.ReadContext()`

File: csharp/src/Ice/Protocol.cs [hunk 2]
```diff
@@ -158,6 +159,45 @@ public static Ice.OutputStream CreateFailureResponseFrame(System.Exception excep
...
+            while (sz-- > 0)
+            {
+                string first = requestFrame.ReadString();
+                string second = requestFrame.ReadString();
+                context[first] = second;
+            }
+
+            var current = new Ice.Current(adapter, identity, facet, operation, (Ice.OperationMode)mode, context,
```
Comment (pepone) [inline csharp/src/Ice/Protocol.cs]: directly return current no need to create this reference

File: csharp/src/Ice/IObject.cs [hunk 3]
```diff
@@ -163,10 +163,10 @@ private static Ice.OutputStream WriteParamEncaps(byte[]? v, bool ok, Current cur
                 return Protocol.CreateEmptyResponseFrame(current);
             }
 
-            if (!ok)
-            {
-                current.DispatchObserver?.UserException();
-            }
+            // if (!ok)
```
Comment (pepone) [inline csharp/src/Ice/IObject.cs]: why is this commented out?

File: csharp/src/Ice/Protocol.cs (hunk 1 above)
Comment (bernardnormier) [reply csharp/src/Ice/Protocol.cs]: It should, yes (I didn't rewrite it, just moved it there). It's possible we could ever use a helper Slice struct here.

File: csharp/src/Ice/IObject.cs (hunk 3 above)
Comment (bernardnormier) [reply csharp/src/Ice/IObject.cs]: Current no longer carries the dispatch observer and the BlobjectAsync will be removed entirely very soon.

File: csharp/src/Ice/Protocol.cs (hunk 1 above)
Comment (bernardnormier) [reply csharp/src/Ice/Protocol.cs]: Fixed!

File: csharp/src/Ice/Protocol.cs (hunk 2 above)
Comment (bernardnormier) [reply csharp/src/Ice/Protocol.cs]: Fixed

**Pull request merged**